Scrape Devpost for sponsor challenge information

naveenr414 opened the pull request:
Created devpost scraper, so that we can get all the challenges and companies from devpost.

naveenr414 committed:
Changed gitignore and updated devpost

naveenr414 committed:
Updated server.py with tab

thisistimothychen commented on api/server.py:
```diff
@@ -14,6 +14,8 @@
 import datetime
 import os
 from seed_db import *
+import config
```
I don't think we need this line - let's remove it so it doesn't get merged in :)

thisistimothychen commented on api/server.py:
```diff
@@ -372,6 +374,49 @@ def delete_all_projects():
     # Number of prizes they can choose per challenge
     # ProjectID that won the challenge
 
+@app.route('/api/import-challenges')
```
Let's do `/api/seed-challenges-from-devpost` to make it really descriptive?

thisistimothychen commented on api/server.py:
```diff
@@ -372,6 +374,49 @@ def delete_all_projects():
 
+@app.route('/api/import-challenges')
+def import_challenges():
+    devpost_url = "https://bitcamp2018.devpost.com/submissions"
```
Let's rename to `DEVPOST_ROOT_URL` and move into `config.py` so that future hackathons don't need to dig through this huge `server.py` file to customize their app, but do everything in a centralized place. Once you move this into `config.py`, you should be able to access via `current_app.config['DEVPOST_ROOT_URL']`

thisistimothychen commented on api/server.py:
```diff
@@ -372,6 +374,49 @@ def delete_all_projects():
+
+    prize_list = get_challenges(devpost_url)
+    company_list = []
+    company_names = list(set([prize[1] for prize in prize_list]))
```
Sick!

thisistimothychen commented on api/server.py:
```diff
@@ -372,6 +374,49 @@ def delete_all_projects():
+    for company_name in company_names:
+        challenge_info = [[prize[0],prize[2]] for prize in prize_list
+                         if prize[1] == company_name]
+
```
(nit) Probably can get rid of this extra line

thisistimothychen commented on api/server.py:
```diff
@@ -372,6 +374,49 @@ def delete_all_projects():
+                         if prize[1] == company_name]
+
+
+        #Autogenerate access_code
```
Awesome. Great job finding this piece and adding it in - loving your thoroughness!

thisistimothychen commented on api/server.py:
```diff
@@ -372,6 +374,49 @@ def delete_all_projects():
+            }
+
+        
+        company_list.append({'company_name':company_name,
```
Let's format as:
```
company_list.append({
    'company_name':company_name,
    'access_code':access_code.upper(),
    'challenges':challenges_obj
})
```
for consistency

thisistimothychen commented on api/server.py:
```diff
@@ -372,6 +374,49 @@ def delete_all_projects():
+        company_list.append({'company_name':company_name,
+                             'access_code':access_code.upper(),
+                             'challenges':challenges_obj})
+        companies.insert(company_list[-1])
```
Would using `companies.append` work? Might be more easily understood

naveenr414 replied:
It's a collections object, so I'm not sure if append would work in this case

thisistimothychen commented on api/devpost_scraper.py:
```diff
@@ -0,0 +1,50 @@
+
+    num_winners = []
+
+    #Find the number of winners for each prize
```
Ah shoot, I should've been more clear with this part -- so the # of winners for a project isn't actually what is on Devpost. This number is how many prizes the company intends to give out. So the timeline goes like this: company gives us info (including the number of prizes they want to give out) -> we create the Devpost stuff -> _then_ we import into hackathon-expo-app using your code. Default number of prizes to award is 1, but and otherwise Devpost includes it in the challenge name.

For example, in this image, Ark is giving out 5, 1517 Fund is giving out 2, and Booz Allen is giving out 1.
![image](https://user-images.githubusercontent.com/14797288/52237375-59023380-2897-11e9-9a9e-54e7679a51e4.png)

thisistimothychen replied:
So I guess a better variable name instead of `num_winners` would be something like `number_potential_winners` but let's not worry about renaming - they would end up being the same number anyways :)

thisistimothychen reviewed (changes requested):
Really great job, Naveen! I'm super impressed with your thoroughness and technical abilities. I made some comments, let's get those addressed and we can merge this in! Looking forward to more of your work!

naveenr414 committed:
Made changes to match with pull request comments

thisistimothychen commented on api/server.py:
```diff
@@ -374,11 +373,13 @@ def delete_all_projects():
+    devpost_url = current_app.config['DEVPOST_ROOT_URL']
     companies = mongo.db.companies
 
+    print(type(companies))
```
Do we still need this?

naveenr414 committed:
Got rid of print statement

naveenr414 committed:
Updated how devpost scraper works

naveenr414 committed:
Deleted CSV

thisistimothychen commented:
Noticed that if you have a company that gives 2 challenges, the endpoint only adds a single challenge.
If you check out https://technica2018.devpost.com/, Technica has 2 challenges, but when I tried to scrape, it only seeded `Best "Light the Way" Hack`.

thisistimothychen commented:
TODOs:
- Change `DEVPOST_ROOT_URL` to `https://bitcamp2018.devpost.com/`
- Change `@app.route` to `@app.route('/api/seed-challenges-from-devpost', methods=['POST'])`
- Add `@is_admin` to secure the endpoint

naveenr414 committed:
Added is_admin

thisistimothychen committed:
Fix overwriting error for companies with multiple challenges

ID was getting generated too fast, so hacky datetime-based id generation method
was breaking. Using company AND challenge name in the id now.

thisistimothychen committed:
Merge branch 'master' into devpost_challenges
